feat(web-analytics): add Active Hours Heat Map query and visualization

jabahamondes opened the pull request:
> [!IMPORTANT]
> 👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Problem

This feature enhances the analytics capabilities by providing a visual representation of active hours, allowing for better insights into user engagement patterns.

## Changes

- Update schema.
- Implemented  query runner.
- Added  component for rendering the heat map visualization.
- Created configuration for heat map display, including color settings and aggregation logic.
- Added tests for the query runner and heat map configuration.

## Does this work well for both Cloud and self-hosted?

<!-- Yes / no / it doesn't have an impact. -->

## How did you test this code?

Manual testing and test files
<img width="1196" alt="image" src="https://github.com/user-attachments/assets/a53267c5-8963-4d44-8400-6ce2dc0f806d" />

Commit:
feat(web-analytics): add Active Hours Heat Map query and visualization

- Introduced  and its response schema.
- Implemented  for executing the query.
- Added  component for rendering the heat map visualization.
- Created configuration for heat map display, including color settings and aggregation logic.
- Added tests for the query runner and heat map configuration.

This feature enhances the analytics capabilities by providing a visual representation of active hours, allowing for better insights into user engagement patterns.

rafaeelaudibert commented on frontend/src/lib/utils.tsx:
```diff
@@ -590,7 +634,7 @@ export function humanFriendlyDuration(
         units = [hDisplay, mDisplay, sDisplay].filter(Boolean)
     }
-    return units.slice(0, maxUnits ?? undefined).join(' ')
+    return units.slice(0, maxUnits ?? undefined).join(' ')
```
I've added these whitespace changes to a PR of mine as well, but they're actually intended. This is a non-breaking whitespace, and we should keep it. I wonder what's removing it.

jabahamondes replied:
It was change without intention, I'm reverting it

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -755,7 +756,7 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
         sourceTab: [(s) => [s._sourceTab], (sourceTab: string | null) => sourceTab || SourceTab.CHANNEL],
         deviceTab: [(s) => [s._deviceTab], (deviceTab: string | null) => deviceTab || DeviceTab.DEVICE_TYPE],
         pathTab: [(s) => [s._pathTab], (pathTab: string | null) => pathTab || PathTab.PATH],
-        geographyTab: [(s) => [s._geographyTab], (geographyTab: string | null) => geographyTab || GeographyTab.MAP],
+        geographyTab: [(s) => [s._geographyTab], (geographyTab: string | null) => geographyTab || GeographyTab.MAP || GeographyTab.HEATMAP],
```
This change doesn't make sense, it'll never get to `HEATMAP` because `GeographyTab.MAP` is always truthy

jabahamondes replied:
Thanks, I'll remove it 🫡

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -1695,7 +1696,7 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
                                   colSpanClassName: 'md:col-span-full',
                               },
-                              activeTabId: geographyTab || GeographyTab.MAP,
+                              activeTabId: geographyTab || GeographyTab.MAP || GeographyTab.HEATMAP,
```
Same here

jabahamondes replied:
idem

rafaeelaudibert commented on posthog/hogql_queries/web_analytics/active_hours_heatmap_query_runner.py:
```diff
@@ -0,0 +1,73 @@
+            )
+            GROUP BY day, hour
+            ORDER BY day, hour
+            LIMIT 168
```
Do we need this? It shouldn't ever return more than 168

jabahamondes replied:
It was added because otherwise it was returning less that the desired results, but I'm gonna verify if this would happen on prod

jabahamondes replied:
I think i will still keep it so in localhost we can see everything good 
<img width="1179" alt="image" src="https://github.com/user-attachments/assets/a075a550-f855-4d8e-a507-b21ec95ffccd" />

rafaeelaudibert commented on posthog/hogql_queries/web_analytics/active_hours_heatmap_query_runner.py:
```diff
@@ -0,0 +1,73 @@
+            timings=self.timings,
+            modifiers=self.modifiers,
+        )
+        results: list[ActiveHoursHeatMapResult] = []
+        for row in response.results:
+            results.append(ActiveHoursHeatMapResult(day=int(row[0]), hour=int(row[1]), total=int(row[2])))
```
Much more pythonic

```suggestion
        results = [
        	ActiveHoursHeatMapResult(day=int(row[0]), hour=int(row[1]), total=int(row[2]))
        	for row in response.results
        ]
```

rafaeelaudibert replied:
What if there isn't any events on some days of the week/some hours of the day?
Maybe we should instead iterate over all the posibilities here and guarantee the array always has the 168 options?

jabahamondes replied:
It returns a list of objects like [{day: 1, hour: 20, total: 5}, ...] so in case there are no results it would be an empty list.

Of the frontend we are mapping only the results (and we have a default table that is filled with 0s) .

So there is no problem with this, and I checked already 🫡 :

<img width="1187" alt="image" src="https://github.com/user-attachments/assets/ccae7f98-be72-4e03-859f-52163ab0294d" />

rafaeelaudibert commented on posthog/hogql_queries/web_analytics/test/test_active_hours_heatmap_query_runner.py:
```diff
@@ -0,0 +1,77 @@
+)
+
+
+class TestActiveHoursHeatMapQueryRunner(ClickhouseTestMixin, APIBaseTest):
```
Adding `@snapshot_clickhouse_queries` is always a good call

Commit:
Reverting automatic changes

Commit:
Adding ff and fixing edge case

Commit:
prs comments

jabahamondes committed:
Update posthog/hogql_queries/web_analytics/active_hours_heatmap_query_runner.py

Co-authored-by: Rafael Audibert <32079912+rafaeelaudibert@users.noreply.github.com>

Commit:
refactor: remove unused humanFriendlyLargeNumberMaxChars function and related tests

- Deleted the  function from  and its corresponding tests in .
- Cleaned up  component by removing color constants and replacing them with theme-based colors.
- Updated styles in  for better layout and alignment.
- Adjusted logic in  to utilize dynamic theme colors for heatmap and aggregation cells.
- Removed unnecessary color tests from  and the  constant from .

Commit:
reverting change

Commit:
reverting change

Commit:
overriding test

jabahamondes committed:
Merge branch 'master' into feat/active-hours-web-analytics-tab

github-actions[bot] committed:
Update query snapshots

Commit:
fixing line quality check

Commit:
running prettier command

Commit:
mypy baseline

jabahamondes committed:
Merge branch 'master' into feat/active-hours-web-analytics-tab

Commit:
prettier

Commit:
build schemas

Commit:
simplifying code

Commit:
updating tests

jabahamondes committed:
Merge branch 'master' into feat/active-hours-web-analytics-tab

Commit:
refactor: rename ActiveHoursHeatMapQuery to WebActiveHoursHeatMapQuery and update related components

- Removed ActiveHoursHeatMapQuery and ActiveHoursHeatMapQueryResponse from schema.
- Introduced WebActiveHoursHeatMapQuery and WebActiveHoursHeatMapQueryResponse.
- Updated all references in the codebase to use the new naming convention.
- Added WebActiveHoursHeatMapQueryRunner for handling query execution.
- Updated tests to reflect the new query structure and naming.

jabahamondes committed:
Merge branch 'master' into feat/active-hours-web-analytics-tab

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

Commit:
Fixing naming

jabahamondes committed:
Merge branch 'master' into feat/active-hours-web-analytics-tab

jabahamondes commented on frontend/src/scenes/web-analytics/EventsHeatMap/EventsHeatMap.tsx:
```diff
@@ -0,0 +1,269 @@
+    for (let i = 0; i < DaysAbbreviated.values.length; i++) {
+        matrix[i] = {}
+        for (let x = 0; x < HoursAbbreviated.values.length; x++) {
+            matrix[i][x] = 0
```
This was not really easy to do, because we cannot now easily the dates range, and if we include some logic in the query it gets much more complicated, so I'll skip this for the moment

Commit:
refactor: update references to WebActiveHoursHeatMapQuery and adjust related components

- Renamed ActiveHoursHeatMapResult to WebActiveHoursHeatMapResult in EventsHeatMap component.
- Changed processData function to accept WebActiveHoursHeatMapResult type.
- Updated background color assignment logic for consistency.
- Adjusted test case name for clarity regarding empty results.
- Minor cleanup in test setup for web analytics queries.

jabahamondes committed:
Update frontend/src/scenes/web-analytics/EventsHeatMap/HeatMapCell.tsx

Co-authored-by: greptile-apps[bot] <165735046+greptile-apps[bot]@users.noreply.github.com>

jabahamondes commented on frontend/src/scenes/web-analytics/EventsHeatMap/config.ts:
```diff
@@ -0,0 +1,24 @@
+}
+
+export const HoursAbbreviated: AxisConfig = {
+    values: Array.from({ length: 24 }, (_, i) => String(i)),
```
I'm gonna skip this for now

jabahamondes commented on posthog/hogql_queries/web_analytics/web_active_hours_heatmap_query_runner.py:
```diff
@@ -0,0 +1,74 @@
+            )
+            GROUP BY day, hour
+            ORDER BY day, hour
+            LIMIT 168
```
added

Commit:
Bot comments

jabahamondes commented on frontend/src/queries/Query/Query.tsx:
```diff
@@ -155,6 +157,8 @@ export function Query<Q extends Node>(props: QueryProps<Q>): JSX.Element | null
         component = <WebVitalsPathBreakdown query={query} cachedResults={props.cachedResults} context={queryContext} />
     } else if (isHogQuery(query)) {
         component = <HogDebug query={query} setQuery={setQuery as (query: any) => void} queryKey={String(uniqueKey)} />
+    } else if (isWebActiveHoursHeatMapQuery(query)) {
+        component = <EventsHeatMap query={query} context={queryContext} />
     } else {
```
Not necessary

Commit:
Passing down cachedResults

jabahamondes committed:
Merge branch 'master' into feat/active-hours-web-analytics-tab

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

jabahamondes commented on posthog/hogql_queries/web_analytics/web_active_hours_heatmap_query_runner.py:
```diff
@@ -0,0 +1,74 @@
+            for row in response.results
+        ]
+
+        assert results is not None
```
I'll include in later prs
